### PR TITLE
Integrated model inside the q function

### PR DESCRIPTION
**Pull request by vFf0621**

### CODE REVIEW

File: agents/alm.py [hunk 1]
```diff
@@ -79,11 +76,14 @@ def get_lower_bound(self, state_batch, action_batch):
             z_batch = self.encoder_target(state_batch).sample()
             z_seq, action_seq = self._rollout_evaluation(z_batch, action_batch, std=0.1)
 
+            reward = self.reward(z_seq[:-1], action_seq[:-1])
             kl_reward = self.classifier.get_reward(z_seq[:-1], action_seq[:-1], z_seq[1:])
             discount = self.gamma * torch.ones_like(reward)
-            q_values_1, q_values_2, reward = self.critic(z_seq[-1], action_seq[-1])
+            if self.critic_mode == "model":
+                q_values_1, q_values_2, reward  = self.critic(z_seq[-1], action_seq[-1])
+            else:
```
Comment (ktolnos) [inline agents/alm.py:84]: It is generally a better idea to check all values explicitly and raise errors from the else branch. This way you will not be misled if you make a typo in config

File: agents/alm.py [hunk 2]
```diff
@@ -311,25 +298,31 @@ def _intrinsic_reward_loss(self, z, action_batch, z_next, z_next_prior, log, met
     def update_critic(self, z_batch, action_batch, reward_batch, z_next_batch, discount_batch, std, log, metrics):
         with torch.no_grad():
             next_action_dist = self.actor(z_next_batch, std)
-            next_action_batch = next_action_dist.rsample(clip=self.stddev_clip)
-
+            next_action_batch = next_action_dist.sample(clip=self.stddev_clip)
+
+        if self.critic_mode == 'model':
+            Q1, Q2, R = self.critic(z_batch, action_batch)
+            Q1_, Q2_, R_ = self.critic(z_next_batch, next_action_batch)
```
Comment (ktolnos) [inline agents/alm.py:305]: Do not change the code that uses critic, change only the critic itself. This way you will make less errors.
```
Q1 = Q1 * self.gamma + R
``` 
should be inside the ModelCritic class.

E.g. in this case you use `critic` where the original code used  critic_target, which is an error that you wouldn't make following this advice.

File: agents/alm.py [hunk 3]
```diff
@@ -311,25 +291,30 @@ def _intrinsic_reward_loss(self, z, action_batch, z_next, z_next_prior, log, met
...
+            next_action_batch = next_action_dist.sample(clip=self.stddev_clip)
+        Q1, Q2 = self.critic(z_batch, action_batch)
+        if self.critic_mode == 'model':
+            Q1_, Q2_ = self.critic(z_next_batch, next_action_batch)
+            Q_ = torch.min(Q1_,Q2_)
+        elif self.critic_mode=='std':
             target_Q1, target_Q2 = self.critic_target(z_next_batch, next_action_batch)
             target_V = torch.min(target_Q1, target_Q2)
-            target_Q = reward_batch.unsqueeze(-1) + discount_batch.unsqueeze(-1)*(target_V)
+            Q_ = reward_batch.unsqueeze(-1) + discount_batch.unsqueeze(-1)*(target_V)
```
Comment (ktolnos) [inline agents/alm.py:302]: Please avoid unnecessary renamings, target_Q was better IMO

File: agents/alm.py [hunk 4]
```diff
@@ -342,24 +327,19 @@ def update_actor(self, actor_loss, log, metrics):
...
-        return actor_loss
 
     def _lambda_svg_loss(self, z_seq, action_seq, std, log, metrics):
         with utils.FreezeParameters([self.model, self.reward, self.classifier, self.critic]):
             reward = self.reward(z_seq[:-1], action_seq[:-1])
             kl_reward = self.classifier.get_reward(z_seq[:-1], action_seq[:-1], z_seq[1:].detach())
             discount = self.gamma * torch.ones_like(reward)
-            q_values_1, q_values_2 = self.critic(z_seq, action_seq.detach())
-            q_values = torch.min(q_values_1, q_values_2)
+            if self.critic_mode == "model":
```
Comment (ktolnos) [inline agents/alm.py:336]: This if seems unnecessary

File: cfgs/agent/alm.yaml [hunk 5]
```diff
@@ -1,28 +1,27 @@
...
-seed: 640
+seed: 639
 
 #benchmark
 benchmark: 'gym'
 id: 'HalfCheetah-v2'
 
 #algorithmic config
-rollout: true_model # alm, rollout_correction, 1_step_correction, true_model
+rollout: rollout_correction # alm, rollout_correction, 1_step_correction
```
Comment (ktolnos) [inline cfgs/agent/alm.yaml:11]: Please, use alm for this experiment, I want to see only changes from replacing the critic

File: cfgs/agent/alm.yaml [hunk 6]
```diff
@@ -1,28 +1,27 @@
...
 model_mode: diff # full (alm), diff
```
Comment (ktolnos) [inline cfgs/agent/alm.yaml:12]: full

File: cfgs/agent/alm.yaml [hunk 7]
```diff
@@ -1,28 +1,27 @@
...
 model_min_std: 0.1 # default 0.1
 model_max_std: 10 # default 10
 actor_sequence_retrieval: mean # sample, mean
```
Comment (ktolnos) [inline cfgs/agent/alm.yaml:15]: sample

File: cfgs/agent/alm.yaml [hunk 8]
```diff
@@ -1,28 +1,27 @@
...
-rollout_accuracy_batch_size: 10 # 0 to not log
```
Comment (ktolnos) [inline cfgs/agent/alm.yaml:16]: 0 if you want to improve performance (not a useful metric for this experiment)

File: cfgs/agent/alm.yaml [hunk 9]
```diff
@@ -1,28 +1,27 @@
...
 #data 
 num_train_steps: 500000
 explore_steps: 5000
 max_episode_steps: 1000
 env_buffer_size: 100000
-batch_size: 512
+batch_size: 50
 seq_len: 3
-
+critic_mode: model
```
Comment (ktolnos) [inline cfgs/agent/alm.yaml:24]: leave a comment with possible values please

File: cfgs/agent/alm.yaml [hunk 10]
```diff
@@ -1,28 +1,27 @@
...
 actor_sequence_retrieval: mean # sample, mean
-rollout_accuracy_batch_size: 10 # 0 to not log
 
 #data 
 num_train_steps: 500000
 explore_steps: 5000
 max_episode_steps: 1000
 env_buffer_size: 100000
-batch_size: 512
+batch_size: 50
```
Comment (ktolnos) [inline cfgs/agent/alm.yaml:22]: I believe 512 was default
